### ISSUE TITLE
[stmt.if] p2 The divergence of implementations regarding the check of semantics in the discard statement

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#78]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[stmt.if] p2 says
>  During the instantiation of an enclosing templated entity ([temp.pre]), if the condition is not value-dependent after its instantiation, **the discarded substatement (if any) is not instantiated**.  

Consider this example
````cpp
struct S{};

template<class T>
void fun(T){
      if constexpr (false){
        S s = 0;  // #1
    }
}

int  main(){
   fun(0);
}
````
GCC accepts this example but Clang rejects it.  The condition of the *constexpr if* statement is indeed not value-dependent after its instantiation, the discard substatement is not instantiated, therefore the ill-formed construct at `#1` is not instantiated? Even though
[temp.res.general] p6 says

> The validity of a template may be checked prior to any instantiation. The program is ill-formed, no diagnostic required, if:  
>> - no valid specialization can be generated for a template or a substatement of a constexpr if statement within a template and the template is not instantiated

Even it is not clear whether such *a-a-a* invalid construct can exist in the discard statement in a templated entity?

### suggested resolution

 Presumably, the intent is make the following example well-formed
````cpp
template<class T>
concept is_initialized_from_int = requires{
    T{0};
};

struct S{};

template<class T>
void fun(){
      if constexpr (is_initialized_from_int<T>){
        T s = 0;  // #1
        // S s0 = 0; Clang reports an error 
    }
}

int  main(){
   fun<S>();
}
````
Since the declaration at `#1` references a template parameter, the check of semantic should be deferred until the instantiation context, since the construct is in a discard statement, the check will never be performed. In other words, the check of the semantics for the constructs that do not reference any template parameter still is done in the definition context, and the behavior is regulated by [temp.res.general] p6.

**Comment by jensmaurer** [comment]
The discarded substatement is not instantiated, but [temp.res.general] p6 nonetheless makes this ill-formed, no diagnostic required. Those implementations that diagnose and those that don't are both conforming.

I'm not seeing the intent of your "a-a-a" example; it's another case of [temp.res.general] p6, making this ill-formed, no diagnostic required. MSVC chooses not to issue a diagnostic, other implementations possibly do.

I'm not seeing a defect here.

**Comment by xmh0511** [comment]
> I'm not seeing a defect here.

The defect here is what does "the discarded substatement (if any) is not instantiated" mean? For example:
````cpp
template<class T>
void fun(){
   if constexpr(false){  // always be discard statement
      S s = 0; // no conversion 
  }
}
````
When [temp.res.general] p6 bullet 1 applies to this example, **in any specialization**, the statement is discarded. Does any semantic check need to apply to the construct in the discard statement? Or, the intent of saying that "the discarded substatement (if any) is not instantiated" is to make that any semantic check cannot be performed on the constructs that depend on template parameters, in other words, any construct **in the discard statement** that **does not depend on template parameters** is still restricted by semantics.

**Comment by jensmaurer** [comment]
Yes,  [temp.res.general] p6.4 seems to work better here. Since instantiation replaces appearances of template parameters with their template argument values, obviously nothing is dependent anymore after instantiation. So any mention of "does not depend on a template parameter" obviously refers to the non-instantiated template.

Independently, would it help if  [temp.res.general] p6.1 talked about "and the template <ins>or substatement</ins> is not instantiated", given that we say in [temp.inst] p11 that a constexpr if substatement is independently instantiated?

**Comment by xmh0511** [comment]
Yes. It would be even better to clarify what the verb "instantiate" in this document means.   

> replaces appearances of template parameters with their template argument values

This utterance has multi-times appeared in [temp.deduct] but we say nothing about what "instantiate" exactly means, which causes the meaning of instantiation to be quite abstract/vacuous when we mention it in the rules.

**Comment by jensmaurer** [comment]
We agree we lack a normative definition of "instantiate".

**Comment by xmh0511** [comment]
Incidentally, [temp.res.general] p6 also remains some issues. 

[temp.spec.general] p4 says
> A **specialization** is a **class, variable, function, or class member** that is either instantiated ([temp.inst]) from a templated entity or is an explicit specialization ([temp.expl.spec]) of a templated entity.

However, [temp.res.general] p6 bullet 1 says
> no valid **specialization** can be generated for ... or a substatement of a constexpr if statement within a template

what does **specialization for a substatement of a constexpr if statement** mean? In this document, we just say "a specialization for a member function or static data member of a class template", "a specialization for a class member of a class template", or "a specialization for a template". 

Is it more clear if [temp.res.general] p6 says
> no valid **specialization** can be generated for a template, or no specialization **referred to** in a substatement of a constexpr if statement(if instantiated) can be valid. 


Moreover, consider this example:
````cpp
template<class T> class X {
  void f(T t, int i, char* p) {
    if constexpr(false){   // never instantiated
       X<T>::h(t);      // may be diagnosed even if X​::​f is not instantiated
    }  
  }
};
````
It seems that there is no bullet in [temp.res.general] p6 clearly specifies this example. 

Change [temp.res.general] p6.4 might cover this example
> a hypothetical instantiation of a template immediately following its definition would be ill-formed due to a construct ~that does not depend on a template parameter~

The discard statement is not instantiated, so the construct `X<T>::h(t)` always depends on the template parameters, in any context. However, such a construct is certainly ill-formed since the result of the lookup can be determined in the template definition context regardless of what the template argument value is.

**Comment by jensmaurer** [comment]
> The discard statement is not instantiated, so the construct `X<T>::h(t)` always depends on the template parameters, in any context.

No. `X<T>` names the current instantiation, so the name lookup for `X<T>::h` is not in a dependent context.

**Comment by xmh0511** [comment]
> X<T> names the current instantiation, so the name lookup for X<T>::h is not in a dependent context.

That is to say, `h` is not a dependent name, so [basic.lookup.general] p1 should apply here without any other exception
> Unless otherwise specified, the program is ill-formed if no declarations are found.  

However, the special in this case is, the expression appears in a discard statement, which is not instantiated, and [temp.res.general] p6.4 merely says
> a hypothetical instantiation of a template immediately following its definition would be ill-formed due to a construct **that does not depend on a template parameter**, or

For `X<T>::h(t); `, `h` is not a dependent name but it does depend on a template parameter as per [temp.dep.type] p7, in other words, in any hypothetical instantiation, the discard statement is never instantiated, that is, `X<T>::h(t); ` always depends on a template parameter, so p6.4 does not apply to this case.

**Comment by jensmaurer** [comment]
> For `X<T>::h(t);` , h is not a dependent name but it does depend on a template parameter as per [temp.dep.type] p7,

The entire call `X<T>::h(t);` depends on a template parameter (because the type of `t` is dependent), but the part `X<T>::h` (i.e. the lookup of the function name) does not.

**Comment by xmh0511** [comment]
> The entire call X<T>::h(t); depends on a template parameter (because the type of t is dependent), but the part X<T>::h (i.e. the lookup of the function name) does not.

Yes, so I said in the above comment that the **qualified name** is **not a dependent name** as per [temp.dep.type] p5. Hence, [basic.lookup] p1 can directly work here with no further exceptions(i.e. the name is not dependent). 

> Unless otherwise specified, the program **is ill-formed** if no declarations are found.

So, the comment is wrong here
> X<T>::h();      // may be diagnosed even if X​::​f is not instantiated

It should be diagnosed since it has violated [basic.lookup] p1 which didn't say no diagnosis required.

**Comment by jensmaurer** [comment]
I don't think we require diagnostics for non-instantiated templates, no even for non-dependent constructs therein, because we want to support token-replay implementations. Maybe the words aren''t there for that.

**Comment by xmh0511** [comment]
> I don't think we require diagnostics for non-instantiated templates, no even for non-dependent constructs therein, because we want to support token-replay implementations.

name lookup rules apply uniformly to all names, as per [basic.lookup]. And the case does violate the rule defined in P1 even though its a template, we still say the lookup rule is performed in the **template definition context**, and the result finds nothing and the name is a non-dependent name, so "Unless otherwise specified" cannot apply here, so 
> the program is ill-formed if no declarations are found.

We may want the validity of a template to be regulated by [temp.res.general] p6 and at that time to check whether they violate other rules even though in the definition context it has obviously violated a rule defined in other clauses.  In the status quo, such two wording are contradictory.
